token_examples/factory_token/src/token.rs

```diff
@@ -0,0 +1,423 @@
+}
+
+impl Cbor for MintParams {}
+
+impl BasicToken {
+    pub fn new<BS: _BS>(
+        bs: &BS,
+        name: String,
+        symbol: String,
+        granularity: u64,
+        minter: Option<Address>,
+    ) -> Self {
+        BasicToken { token: TokenState::new(&bs).unwrap(), name, symbol, granularity, minter }
+    }
+
+    pub fn token(&mut self) -> Token<'_, Blockstore, FvmMessenger> {
+        let bs = Blockstore::default();
+        let msg = FvmMessenger::default();
+        Token::wrap(bs, msg, self.granularity, &mut self.token)
+    }
+
+    pub fn load(cid: &Cid) -> Result<Self, RuntimeError> {
+        let bs = Blockstore::default();
+        match bs.get_cbor::<Self>(cid) {
+            Ok(Some(s)) => Ok(s),
+            // TODO: improve on these errors?
+            Ok(None) => Err(RuntimeError::Deserialization("no data found".into())),
+            Err(e) => Err(RuntimeError::Deserialization(e.to_string())),
+        }
+    }
+
+    pub fn save(&self) -> Result<Cid, RuntimeError> {
+        let bs = Blockstore::default();
+        let serialized = fvm_ipld_encoding::to_vec(self)
+            .map_err(|err| RuntimeError::Serialization(err.to_string()))?;
+        let block = Block { codec: DAG_CBOR, data: serialized };
+        bs.put(Code::Blake2b256, &block).map_err(|err| RuntimeError::Serialization(err.to_string()))
+    }
+
+    fn reload(&mut self, initial_cid: &Cid) -> Result<(), RuntimeError> {
+        let new_cid = sdk::sself::root()?;
+        if new_cid != *initial_cid {
+            let new_state = Self::load(&new_cid)?;
+            let _old = std::mem::replace(self, new_state);
+        }
+        Ok(())
+    }
+
+    pub fn mint(&mut self, params: MintParams) -> Result<MintReturn, RuntimeError> {
+        // todo: check that caller matches our minter address
+        let caller = caller_address();
+        // no minter means minting has been permanently disabled
+        let minter = self.minter.ok_or(RuntimeError::MintingDisabled)?;
+        // if there is a minter, check them
+        let msg = FvmMessenger::default();
+        if !msg.same_address(&caller, &minter) {
+            return Err(RuntimeError::AddressNotAuthorized);
+        }
+
+        let mut hook = self.token().mint(
+            &caller,
+            &params.initial_owner,
+            &params.amount,
+            params.operator_data,
+            Default::default(),
+        )?;
+
+        //let cid = self.token().flush()?;
+        let cid = self.save()?;
+        sdk::sself::set_root(&cid).unwrap();
+
+        let hook_ret = hook.call(self.token().msg())?;
+
+        self.reload(&cid)?;
+        let ret = self.token().mint_return(hook_ret)?;
+
+        Ok(ret)
+    }
+
+    /// Permanently disable minting
+    /// Only the authorised mint operator can do this
+    ///
+    /// TODO: should authorised minter be separate to a token 'owner' ?
+    /// also, should it be a single address, a list of addresses, or some other means?
```

Comment:
Thoughts? I did what was easy (the single operator for minting) but not sure if it's the best approach

Reply:
It's now a single `ActorID`, as suggested in another comment